tfhe/src/core_crypto/algorithms/lwe_encryption.rs

```diff
@@ -57,6 +112,51 @@ pub fn fill_lwe_mask_and_body_for_encryption<Scalar, KeyCont, OutputCont, Gen>(
     ));
 }
 
+pub fn fill_lwe_mask_and_body_for_encryption_other_mod<Scalar, KeyCont, OutputCont, Gen>(
+    lwe_secret_key: &LweSecretKey<KeyCont>,
+    output_mask: &mut LweMask<OutputCont>,
+    output_body: &mut LweBodyRefMut<Scalar>,
+    encoded: Plaintext<Scalar>,
+    noise_parameters: impl DispersionParameter,
+    generator: &mut EncryptionRandomGenerator<Gen>,
+) where
+    Scalar: UnsignedTorus,
+    KeyCont: Container<Element = Scalar>,
+    OutputCont: ContainerMut<Element = Scalar>,
+    Gen: ByteRandomGenerator,
+{
+    assert_eq!(
+        output_mask.ciphertext_modulus(),
+        output_body.ciphertext_modulus(),
+        "Mismatched moduli between mask ({:?}) and body ({:?})",
+        output_mask.ciphertext_modulus(),
+        output_body.ciphertext_modulus()
+    );
+
+    let ciphertext_modulus = output_mask.ciphertext_modulus();
+
+    assert!(!ciphertext_modulus.is_compatible_with_native_modulus());
+
+    generator.fill_slice_with_random_mask_custom_mod(output_mask.as_mut(), ciphertext_modulus);
+
+    let ciphertext_modulus_as_scalar: Scalar = ciphertext_modulus.get_custom_modulus().cast_into();
+
+    // generate an error from the normal distribution described by std_dev
+    *output_body.data = generator.random_noise_custom_mod(noise_parameters, ciphertext_modulus);
+    *output_body.data =
+        (*output_body.data).wrapping_add_custom_mod(encoded.0, ciphertext_modulus_as_scalar);
+
+    // compute the multisum between the secret key and the mask
+    *output_body.data = (*output_body.data).wrapping_add_custom_mod(
+        slice_wrapping_dot_product_custom_mod(
+            output_mask.as_ref(),
+            lwe_secret_key.as_ref(),
+            ciphertext_modulus_as_scalar,
+        ),
+        ciphertext_modulus_as_scalar,
+    );
```

Comment:
what do you mean ?

Reply:
```
let random_noise = ...
let dot_product = ...
*output_body.data = dot_product.wrapping_add(random_noise).wrapping_add(encoded)
```

Reply:
re reading the code I think my idea was : have CSPRNG actions close together (having the CSPRNG code already "hot" from the mask generation)
never store the noise in a temporary to avoid leaking the value if possible, though this may not matter much

I'll see if I can achieve this and have something less complicated to read

Reply:
not sure having the dot product in a temporary is a good idea either...

Reply:
does this look ok ? keeping noise generation close to mask generation
```rust
    // Store delta * m + error, keeping noise generation close to mask generation as that code path
    // will already be hot
    *output_body.data = encoded
        .0
        // add the noise: an error generated from the normal distribution described by std_dev
        .wrapping_add(generator.random_noise_custom_mod(noise_parameters, ciphertext_modulus));

    // compute the multisum between the secret key and the mask
    let mask_key_dot_product = (*output_body.data).wrapping_add(slice_wrapping_dot_product(
        output_mask.as_ref(),
        lwe_secret_key.as_ref(),
    ));

    // Store sum(ai * si) + delta * m + e in the body
    *output_body.data = (*output_body.data).wrapping_add(mask_key_dot_product);
    // (power of 2 encoding omitted but managed as the last step)
```

Reply:
I don´t see how not explicitly storing temporaries in local variables would prevent anything
In LLVM IR, these temporaries will be represented explicitly

Reply:
good point